### PR TITLE
Extract RequestHandler trait

### PR DESCRIPTION
**Pull request by breeswish**
Extracted from https://github.com/pingcap/tikv/pull/2896

### CODE REVIEW

File: src/coprocessor/checksum.rs [hunk 1]
```diff
@@ -112,6 +88,38 @@ impl ChecksumContext {
...
+        resp.set_total_kvs(total_kvs);
+        resp.set_total_bytes(total_bytes as u64);
+        let data = box_try!(resp.write_to_bytes());
+
+        let mut resp = Response::new();
+        resp.set_data(data);
+        Ok(resp)
+    }
+
+    fn collect_metrics_into(&mut self, metrics: &mut self::dag::executor::ExecutorMetrics) {
```
[the remark this answers is not in the document]
Comment (breeswish) [reply src/coprocessor/checksum.rs]: you are right!

File: src/coprocessor/endpoint.rs [hunk 2]
```diff
@@ -185,29 +186,31 @@ impl Host {
...
+                let mut ctx = ChecksumContext::new(checksum, ranges, snap, &req_ctx).unwrap();
                 let do_request = move |_| {
                     tracker.record_wait();
-                    let mut resp = ctx.handle_request(&mut exec_metrics).unwrap_or_else(|e| {
+                    let mut resp = ctx.handle_request().unwrap_or_else(|e| {
                         let mut metrics = tracker.get_basic_metrics();
                         err_resp(e, &mut metrics)
                     });
+                    let mut exec_metrics = ExecutorMetrics::default();
+                    ctx.collect_metrics_into(&mut exec_metrics);
```
Comment (AndreMouche) [inline src/coprocessor/endpoint.rs]: Seems L206~218 repeated in all Unary request, could we merge it now?

Comment (breeswish) [reply src/coprocessor/endpoint.rs]: that's exactly what we will do in #2896

**Pull request merged**